src/irr/asset/CGraphicsPipelineLoaderMTL.cpp

```diff
@@ -639,10 +639,43 @@ SAssetBundle CGraphicsPipelineLoaderMTL::loadAsset(io::IReadFile* _file, const I
             }
         }
 
+        constexpr size_t DS1_METADATA_ENTRY_CNT = 3ull;
+        core::smart_refctd_dynamic_array<IPipelineMetadata::ShaderInputSemantic> shaderInputsMetadata = core::make_refctd_dynamic_array<decltype(shaderInputsMetadata)>(ds3 ? ds3->getTotalDescriptorCount()+DS1_METADATA_ENTRY_CNT : DS1_METADATA_ENTRY_CNT);
+        if (ds3)
+        {
+            ICPUDescriptorSetLayout* ds3layout = ds3->getLayout();
+            for (uint32_t i = 0u; i < shaderInputsMetadata->size()-DS1_METADATA_ENTRY_CNT; ++i)
+            {
+                auto& semantic = (*shaderInputsMetadata)[i];
+                semantic.descriptorSection.type = IPipelineMetadata::ShaderInput::ET_COMBINED_IMAGE_SAMPLER;
+                semantic.descriptorSection.combinedImageSampler.binding = ds3layout->getBindings().begin()[i].binding;
+                semantic.descriptorSection.combinedImageSampler.set = 3u;
+                semantic.descriptorSection.combinedImageSampler.viewType = (semantic.descriptorSection.combinedImageSampler.binding==CMTLPipelineMetadata::EMP_REFL_POSX) ? IImageView<ICPUImage>::ET_CUBE_MAP : IImageView<ICPUImage>::ET_2D;
+                semantic.type = (semantic.descriptorSection.combinedImageSampler.binding == CMTLPipelineMetadata::EMP_REFL_POSX) ? IPipelineMetadata::ECSI_ENVIRONMENT_CUBEMAP : IPipelineMetadata::ECSI_COUNT;
```

Comment:
I think your condition is a typo here.

Reply:
Also add more `ECSI_` enums for bumpmaps, derivative maps, normal maps, albedo maps, etc.

Reply:
Also why is a cubemap in descriptor set 3?
Should be at least 2 or 1.

Reply:
btw the already provided inputs (push constants, material textures) shouldn't be in this metadata.

Reply:
>I think your condition is a typo here.

no, it's ok. why?

>Also why is a cubemap in descriptor set 3? Should be at least 2 or 1.

what why?

>btw the already provided inputs (push constants, material textures) shouldn't be in this metadata.

oh, ok, fine

Reply:
>what why?

so should i have path to refl map in metadata instead + ECSI_ entry?

Reply:
> no, it's ok. why?

Oh its just a badly named enum

> what why?

Does MTL provide the cubemap or does the user provide the cubemap?

Reply:
MTL provides cubemap

> Oh its just a badly named enum

a little, `EMP_REFL_POSX`, `EMP_REFL_NEGX`, etc.. are for loading images (see `CGraphicsPipelineLoaderMTL::readTexture()`), then resulting cubemap image view goes to index `[EMP_REFL_POSX]` of array resulting from `CGraphicsPipelineLoaderMTL::loadImages()`

Reply:
great then we don't need the cubemap in the pipeline metadata

That standardised metadata is only for things we need to provide to make the shader work (not to change it a bit).

Reply:
well, if u want it to be in set 1 or 2 them sure we do

Reply:
well since MTL provides the environment map, it can be in set 3